### PR TITLE
Add the extra load for allocatable

### PR DESCRIPTION
**Pull request by certik**
Test added.

Towards https://github.com/lfortran/lfortran/issues/1874.

### CODE REVIEW

**Comment by certik** [comment]
Currently it segfaults at runtime.

**Comment by certik** [comment]
TODO: just add an LLVM test, since it seems to generate LLVM code that passes verify. That way we can at least compile fastgpt to a binary.

**Review by czgdp1807 — APPROVED** [review]

File: integration_tests/arrays_25.f90 [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+type :: model_t
+    integer, allocatable :: decoder_idx(:)
+end type
+type(model_t) :: m
+allocate(m%decoder_idx(10))
+m%decoder_idx(:) = 0
+m%decoder_idx(5) = 5
+print *, m%decoder_idx
+! Does not compile yet:
+!call decode(m%decoder_idx)
```
Comment (certik) [inline integration_tests/arrays_25.f90:12]: Uncommenting this one fails to compile. Also the generated code segfaults (even with the commented out value).

This PR is not enough for fastGPT, there the LLVM code still fails to verify.

I think we have several bugs there. This PR shows where they are at least.

@czgdp1807 do you know what the issue could be here? You have more experience with the backend how these arrays work. I noticed the following so far: some arrays are allocatable, but only a subset of those fail with the pointer issue. I have no idea what is going on.

**Comment by certik** [comment]
The change in this PR is likely incorrect. It gets half of one test to a binary, as opposed to failing in ASR->LLVM. But the binary segfaults.

I think it's worth fixing this properly to generate a working binary for the full test (not just half). After that I'll run it via fastgpt and generate the next bug report.

File: integration_tests/arrays_25.f90 (hunk 1 above)
Comment (czgdp1807) [reply integration_tests/arrays_25.f90:12]: I will investigate and push the fix tomorrow.

Comment (certik) [reply integration_tests/arrays_25.f90:12]: Thanks! I really appreciate it.

Comment (Smit-create) [reply integration_tests/arrays_25.f90:12]: Uncommenting these lines gives the following error:
```
code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %call_arg_value = alloca %array*, align 8
 i32*  call void @decode_idx(%array** %call_arg_value, i32* %call_arg_value4, i32* %call_arg_value5)
```
`decode_idx` is defined as:
```
define void @decode_idx(i32* %idx, i32* %__1idx, i32* %__2idx)
```

This means `idx` should be an integer pointer, where as we are passing `%array = type { i32*, i32, %dimension_descriptor*, i1, i32 }`.

Therefore we need to somehow load the pointer to the data from `%array`.

Comment (Smit-create) [reply integration_tests/arrays_25.f90:12]: It's confusing to me that how `decode` considers integer arrays by pointers and the main program uses an array descriptor.

Comment (certik) [reply integration_tests/arrays_25.f90:12]: Yes, this is related to physical types that @czgdp1807 is working on, there will be an explicit cast, then it should fix this issue.

Comment (certik) [reply integration_tests/arrays_25.f90:12]: I talked with Gagandeep, let's wait on this issue until physical types are done, and then we'll fix later.

**Comment by certik** [comment]
Once #1813 is in, the following diff makes the test compile to LLVM:
```diff
diff --git a/src/libasr/codegen/asr_to_llvm.cpp b/src/libasr/codegen/asr_to_llvm.cpp
index 6fa2def0c..b6cdbbabc 100644
--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -8826,6 +8826,9 @@ public:
         ASR::array_physical_typeType physical_type = ASRUtils::extract_physical_type(x_mv_type);
         switch( physical_type ) {
             case ASR::array_physical_typeType::DescriptorArray: {
+                if (ASR::is_a<ASR::Allocatable_t>(*x_mv_type) && !(llvm_arg1->getType()->getContainedType(0)->isStructTy())) {
+                    llvm_arg1 = LLVM::CreateLoad(*builder, llvm_arg1);
+                }
                 llvm::Value* dim_des_val = arr_descr->get_pointer_to_dimension_descriptor_array(llvm_arg1);
                 llvm::Value* const_1 = llvm::ConstantInt::get(context, llvm::APInt(32, 1));
                 dim_val = builder->CreateSub(dim_val, const_1);
```
But the resulting binary segfaults. So this is still incorrect (and the diff is hacking anyway, we need to determine this strictly from the ASR itself).

**Comment by certik** [comment]
This is the problematic ASR:
```
(ArrayBound
                                (StructInstanceMember
                                    (Var 2 m)
                                    2 1_model_t_decoder_idx
                                    (Allocatable
                                        (Array
                                            (Integer 4)
                                            [(()
                                            ())]
                                            DescriptorArray
                                        )
                                    )
                                    ()
                                )
                                (IntegerConstant 1 (Integer 4))
                                (Integer 4)
                                UBound
                                ()
                            )
```
The `1_model_t_decoder_idx` is:
```
                            1_model_t_decoder_idx:
                                (ExternalSymbol
                                    2
                                    1_model_t_decoder_idx
                                    3 decoder_idx
                                    model_t
                                    []
                                    decoder_idx
                                    Public
                                ),
```
Which seems like a bug, since there is no module.

**Comment by certik** [comment]
In the lastest `main` (the physical types are now merged), the issue is still exactly as described in the last two comments.

**Comment by certik** [comment]
Maybe one way to make progress on this is what you did --- we have a PR that generates LLVM IR and a binary. The binary segfaults, but I think that's still better than failing even generating LLVM. 

Now the next step is to figure out why the LLVM code is wrong, probably by looking at it and debugging it.

**Comment by Smit-create** [comment]
> Which seems like a bug, since there is no module named model_t.

I think they are handled that way considering that `model_t` is a module (or a node having a symbol table from which we can import). So ExternalSymbol's `a_module_name` doesn't necessarily have to be a module?

https://github.com/lfortran/lfortran/blob/e6bf157813f0eeee9b7c01f6391a79c520cf8a04/src/lfortran/semantics/ast_common_visitor.h#L1064-L1070

**Comment by Smit-create** [comment]
This is the minimal reproducible example that fails:
```
program b
    implicit none
    type :: model_t
        integer, allocatable :: decoder_idx(:)
    end type
    type(model_t) :: m
    allocate(m%decoder_idx(10))

end program
```

This shows that it has something to do with allocating the struct allocatable members.

**Comment by Smit-create** [comment]
I investigated the LLVM code(of https://github.com/lfortran/lfortran/pull/1875#issuecomment-1619479711), and think that we have the following issue in LLVM:
```
define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_set_argv(i32 %0, i8** %1)
  %m = alloca %model_t, align 8
  %2 = getelementptr %model_t, %model_t* %m, i32 0, i32 0
  store %array* null, %array** %2, align 8
```

We have a null value for the array, but now since it is a struct, it must be allocated while initializing the struct `model_t`. I will check if this is the main issue.

**Comment by Smit-create** [comment]
@certik this was the exact issue: https://github.com/lfortran/lfortran/pull/1875#issuecomment-1619496427 and is fixed now. The test you added works perfectly with LLVM.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -8715,6 +8729,10 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                     LLVM::is_llvm_pointer(*ASRUtils::expr_type(x.m_v));
         visit_expr_wrapper(x.m_v);
         ptr_loads = ptr_loads_copy;
+        bool is_pointer_array = tmp->getType()->getContainedType(0)->isPointerTy();
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:8732]: We should figure this out from ASR only, but we can do this later.

**Comment by certik** [comment]
@Smit-create great job!!

You are right, the issue was there. I made a comment above of a smaller issue that we can fix later.

The ExternalSymbol should only be used for other modules, not for symbol access in the same module --- I think that might even fail with serialization / deserialization. We'll fix it later.

Thank you again for fixing this, I couldn't figure it out.

**Pull request merged**